Sema: Remove ConstraintKind::OneWayBindParam and ConstraintKind::OneWayEqual

slavapestov opened the pull request.

slavapestov committed:
Sema: Don't generate OneWayEqual constraints for pattern bindings

slavapestov committed:
Sema: Remove OneWayExpr and Builtin.one_way

slavapestov committed:
Sema: Remove ConstraintKind::OneWayEqual

slavapestov committed:
Sema: Remove one-way constraint handling from computeConnectedComponents()

slavapestov committed:
Sema: Remove DependentComponentSplitterStep

slavapestov committed:
Sema: Simplify depthFirstSearch() a bit now that it only has one caller

xedin commented:
@swift-ci please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

xedin commented:
For posterity.

This might cause a difference in behavior for things like:
```
for (q, a) in [("ultimate question", 42)] where a == 42 {
}
```

because `a` is going to be inferred together with `==` as part of the where clause. This is not the case for closure contexts where `where` clause is always separate from the pattern.

xedin commented:
I think we can follow-up with a small refactoring of `TypeChecker::typeCheckForEachPreamble` to do what CSSyntacticElement does with `where` clause.

xedin committed:
[TypeChecker] Type-check `where` clauses for for-in statements separately

Instead of using `one-way` constraints, just like in closure contexts
for-in statements should type-check their `where` clauses separately.

This also unifies and simplifies for-in preamble handling in the
solver.

xedin committed:
[CSSimplify] Retry member lookup on an insufficiently resolved optional base in diagnostic mode

Re-introduce unsolved member constraint when optional object is
a type variable or member until it's bound, otherwise it's impossible
to tell whether unwrapped base would have a member or not.

AnthonyLatsis reviewed (approved):
(Last 2 commits) Neat!

xedin committed:
[CSGen] forcing an optional should prefer subtyping only if it's applied to a member

`TVO_PrefersSubtypeBinding` in this case is intended to support
references to Objective-C imported members that can be force
unwrapped before being used i.e. accessing an overloaded method
on `AnyObject` should always prefer an overload that returns
a subtype:

```
@objc class A: NSObject {
    @objc func test() -> A { }
    @objc func test() -> B { }
}

func test(obj: AnyObject) {
  obj.test!()
}
```

where `B` is a subtype of `A` should result in selection of
`() -> B`.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility debug

xedin commented:
Debug config failures are unrelated to these changes, main fails the same way - https://ci.swift.org/job/swift-main-source-compat-suite-debug/989/

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility
